### ISSUE TITLE
Suddenly incorrectly transcoding 4k x265 HEVC files on one TV

### ISSUE THREAD

**Issue by WavesOfSalt** [jellyfin-archive/jellyfin-roku-legacy#722]
**Describe The Bug**

4k x265 files stopped trasncoding correctly for a TV that only understands h264 at a maximum of 1080p, only showing a black screenw ith no sound. It's a Roku TV, so no options during playback. 720p and 1080p files encoded with h265 play back fine, so it seems to be that they're not being downscaled anymore. The actual resolution being delivered does not show in transcoding info.

**Media Information**

Entire log directory missing in Windows build or location not documented for Windows. Logging.json does not contain a save directory. Ffmpeg not installed as far as I know.

**Screenshots**

![image](https://user-images.githubusercontent.com/65592397/181111441-fdc29f47-3c9d-4ef2-86e0-8ce37100721f.png)

**System (please complete the following information):**
 - Platform: Roku
 - Jellyfin Version: 10.8.1

**Comment by cewert** [comment]
@WavesOfSalt what version of the roku jellyfin app are you using? v1.5.0 is the most recent version and was just released

**Comment by WavesOfSalt** [comment]
It wouldn't upgrade automatically. I uninstalled and reinstalled to update, but v1.5 gives the same issue.

**Comment by WavesOfSalt** [comment]
In addition. the previous version was 1.4. Is any more info needed? I tried unstable builds of Jellyfin server, and the result was the same.

**Comment by WavesOfSalt** [comment]
Could someone remove the info-needed label? There is no known missing info.

**Comment by ApexArray** [comment]
I am also experiencing this issue.

Jellyfin server 10.8.8 (docker)
Jellyfin Roku client 1.6
Roku 3800x
Vizio 1080p TV

Tested with every sane transcoding option. Only occurs with HEVC 4k content. HEVC 1080p content is fine.

**Comment by ApexArray** [comment]
It looks like this issue has gone under the radar, so I decided to take a stab at it. I dug into the source code and documentation today and I think I found the issue. The Roku app is not reporting the screen resolution info to Jellyfin, so it defaults to sending the video at 4k, which cannot play on my 1080p TV.

For reference the jellyfin-web repo is [reporting the max screen width](https://github.com/jellyfin/jellyfin-web/blob/ed376a263aa81436ae0e6f36f6db4ff9e7685208/src/scripts/browserDeviceProfile.js#L957-L962) with the device capabilities.

I called Roku's [GetSupportedGraphicsResolutions()](https://developer.roku.com/en-gb/docs/references/brightscript/interfaces/ifdeviceinfo.md#getsupportedgraphicsresolutions-as-object) and inserted it into [deviceCapabilities.brs](https://github.com/jellyfin/jellyfin-roku/blob/unstable/source/utils/deviceCapabilities.brs) and that did the trick. ! I can finally watch my 4k video files on my 1080p TV!

I know the general consensus is to avoid transcoding 4k content, but I'm one of the stubborn folks that only wants to store one file for each of my videos and have it work on all my devices, while still getting the max resolution for each respective device. 

I will try to have a PR ready soon.

**Comment by sevenrats** [comment]
I have my doubts that this is the correct way to address this problem. Please correct me if I'm wrong, but, for example, it appears that web only uses this metric on XBox: https://github.com/jellyfin/jellyfin-web/blob/ed376a263aa81436ae0e6f36f6db4ff9e7685208/src/scripts/browserDeviceProfile.js#L381
I kinda suspect that this is one of those sneaky HDR bugs.  Would you be willing to post ffprobes of some videos that have this problem?

**Comment by ApexArray** [comment]
Actually, I did some more testing and it wasn’t the max width that fixed it. I think I these are the lines that fixed it.

```
{
    "Condition": "LessThanEqual",
    "Property": "VideoBitrate",
    "Value": "12000000",
    IsRequired: true
}
```

I was testing multiple things at once last night and didn’t realize I still had that enabled. I just tested again with no max width and confirmed the video only plays if the max bit rate is configured (12Mbps in this case in stead of 120Mbps)

This device happens to be on a wireless connection, and I don’t think it can reach the default 120Mbps. So I guess it was failing because the default bit rate is too high for my connection?

I’ll test more and see if I can find a way to gracefully fallback to a lower bit rate instead of hard capping it.

**Comment by ApexArray** [comment]
I set a breakpoint in the debugger and found that the device is actually running out of memory:
![image](https://user-images.githubusercontent.com/91706106/208316802-3f43974c-9c28-4c04-b874-a26f1699c020.png)

Roku provides a function [EnableLowGeneralMemoryEvent](https://developer.roku.com/en-gb/docs/references/brightscript/interfaces/ifdeviceinfo.md#enablelowgeneralmemoryeventenabled-as-boolean-as-dynamic) to subscribe to low memory events. And the [roDeviceInfoEvent](https://developer.roku.com/en-gb/docs/references/brightscript/events/rodeviceinfoevent.md) page suggests how to deal with "low" or "critical" memory events:

>The application may voluntarily free up memory by invalidating references to objects (e.g. release ContentNodes held in a cache, release offscreen renderable nodes, etc.).

If there's enough wiggle room to free up resources on low memory events, then that's probably the ideal option. But this is my first experience with Roku development and I'm not familiar with that process.

Alternatively, I'm happy to write a function that lowers the max bit rate for [hardware models with 512MB RAM](https://developer.roku.com/en-gb/docs/specs/hardware.md).

@sevenrats, thoughts?

**Comment by Shadowghost** [comment]
If this is bitrate specific it might be fixed server side by https://github.com/jellyfin/jellyfin/pull/8850 otherwise the item bitrate might be wrong witin the media file, the connection too slow or the Roku just can't handle the bitrate.

**Comment by ApexArray** [comment]
@Shadowghost you might have missed my last comment. High bitrates are causing the device to run out of memory.

**Comment by ApexArray** [comment]
Okay, I think I finally got to the crux of the issue. According to [this chart](https://developer.roku.com/en-gb/docs/specs/media/streaming-specifications.md#supported-video-codecs), Roku devices only support h264 "up to 10Mpbs", and h265 "up to 40Mpbs". with some wiggle room for "Peak Video Bit Rate".

To test this, I set the bitrate cap to 24000000 (24Mbps) and saw that it also failed, even though it's not a terribly high bitrate.

I also found [this thread](https://community.roku.com/t5/Solving-playback-issues/33-and-CRASH-time-on-4800r-models/m-p/772892/highlight/true#M38087) where others are encountering similar issues with high bitrate content between Roku and Plex.

I suggest setting the respective max bitrates in each of the `CodecProfiles`. I will close my PR and submit a new one.
